activejob/lib/active_job/concurrency.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+module ActiveJob
+  module Concurrency
+    extend ActiveSupport::Concern
+    extend ActiveSupport::Autoload
+
+    autoload :Limit
+
+    included do
+      class_attribute :_concurrency_limit, instance_accessor: false
+      class_attribute :_concurrency_keys, instance_accessor: false
+      class_attribute :_concurrency_timeout, instance_accessor: false
+    end
+
+    module ClassMethods
+      def concurrency_limit
+        self._concurrency_limit
+      end
+
+      def concurrency_keys
+        self._concurrency_keys
+      end
+
+      def concurrency_timeout
+        self._concurrency_timeout
+      end
+
+      def concurrency(limit:, keys: [], timeout: 120)
+        self._concurrency_limit = limit
+        self._concurrency_keys = keys
+        self._concurrency_timeout = timeout
+      end
+    end
+
+    attr_accessor :concurrency_limit
+
+    attr_accessor :concurrency_keys
+
+    attr_accessor :concurrency_timeout
+
+    def serialize
+      super.merge(
+        "concurrency_limit"   => concurrency_limit,
+        "concurrency_keys"    => concurrency_keys,
+        "concurrency_timeout" => concurrency_timeout
+      )
+    end
+
+    def deserialize(job_data)
+      super
+      self.concurrency_limit   = job_data["concurrency_limit"]
+      self.concurrency_keys    = job_data["concurrency_keys"]
+      self.concurrency_timeout = job_data["concurrency_timeout"]
+    end
+
+    def concurrency_limit
+      self.class.concurrency_limit
+    end
+
+    def concurrency_keys
+      self.class.concurrency_keys
+    end
+
+    def concurrency_timeout
+      self.class.concurrency_timeout
+    end
+
+    def concurrency_limit_instance
+      @concurrency_limit_instance ||= Limit.new(self.class.concurrency_limit)
+    end
+
+    def concurrency_reached?
+      return false unless concurrency_limit_instance
+
+      if concurrency_limit_instance.locking? || concurrency_limit_instance.enqueue_limit?
+        self.class.queue_adapter.concurrency_reached?(self)
+      end
+    end
+
+    def concurrency_key
+      keys = self.class.concurrency_keys
+      return unless keys
+
+      "#{self.class}:#{arguments[0].dig(*keys)}"
+    end
```

Comment:
I believe this might cause some issues down the line, because not all jobs enqueue jobs where arguments is a hash. 

I belive we can overcome this by changing this API to use a block rather than a set of keys. Currently, at Shopify we use the set of keys. Why? So developers do not get extra fancy with the block logic and it also allow us to raise an `Exception` if the specify a key that is no longer in the job params.

This is how the API would look like using a block:

```ruby
class MyJob < ActiveJob::Base
  concurrency(limit: 1, key: -> { |job|
    job.arguments[0]
  })
end
```

I still believe we have to add the `job_class` as part of the queue to guarantee the uniqueness between jobs. Two jobs could enqueue different jobs with the same params.